README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Mcomp: Data from the M-Competitions

[![Travis-CI Build
Status](https://travis-ci.org/robjhyndman/Mcomp.svg?branch=master)](https://travis-ci.org/robjhyndman/Mcomp)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/Mcomp)](https://cran.r-project.org/package=Mcomp)
[![Downloads](http://cranlogs.r-pkg.org/badges/Mcomp)](https://cran.r-project.org/package=Mcomp)
[![Licence](https://img.shields.io/badge/licence-GPL--3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0.en.html)

The R package *Mcomp* contains the 1001 time series from the
M-competition (Makridakis et al. 1982) and the 3003 time series from the
IJF-M3 competition (Makridakis and Hibon, 2000). See also the
[tscompdata package](https://github.com/robjhyndman/tscompdata).

## Installation

You can install the **stable** version on [R
CRAN](https://cran.r-project.org/package=Mcomp).

``` r
install.packages('Mcomp')
```

You can install the **development** version from
[Github](https://github.com/robjhyndman/Mcomp)

``` r
# install.packages("devtools")
devtools::install_github("robjhyndman/Mcomp")
```

## Usage

``` r
library(Mcomp)
#> Loading required package: forecast
library(ggplot2)
M1
#> M-Competition data: 1001 time series 
#> 
#>            Type of data
#> Period      DEMOGR INDUST MACRO1 MACRO2 MICRO1 MICRO2 MICRO3 Total
#>   MONTHLY       75    183     64     92     10     89    104   617
#>   QUARTERLY     39     18     45     59      5     21     16   203
#>   YEARLY        30     35     30     29     16     29     12   181
#>   Total        144    236    139    180     31    139    132  1001
autoplot(M1$YAF2)
```

![](man/figures/README-usage-1.png)<!-- -->

``` r
subset(M1,"monthly")
#> M-Competition data: 617 MONTHLY time series
#> 
#>          Type of data
#> Period    DEMOGR INDUST MACRO1 MACRO2 MICRO1 MICRO2 MICRO3
#>   MONTHLY     75    183     64     92     10     89    104
```

## Sources

[Makridakis, S., A. Andersen, R. Carbone, R. Fildes, M. Hibon, R.
Lewandowski, J. Newton, E. Parzen, and R. Winkler (1982) The accuracy of
extrapolation (time series) methods: results of a forecasting
competition. *Journal of Forecasting*, **1**,
111–153.](http://doi.org/10.1002/for.3980010202)

[Makridakis and Hibon (2000) The M3-competition: results, conclusions
and implications. *International Journal of Forecasting*, **16**,
451-476.](https://doi.org/10.1016/S0169-2070\(00\)00057-1)

## License

This package is free and open source software, licensed under GPL-3


NEWS.md
## Version 2.9 (??)
  * Added some vignettes

## Version 2.8 (19 June 2018)
  * Fixed S3 generic consistency for autoplot.Mdata

## Version 2.7 (2 December 2017)
  * Converted to roxygen
  * Added pkgdown site
  * Improved subsetting via [

## Version 2.6 (23 February 2017)
  * Fixed error with time series not starting at correct fractional year
  * Added autoplot.Mcomp()
  * Removed package dependencies
  * Corrected some labels

## Version 2.05 (15 June 2013)
  * Fixed an error in time series N2851. The last value was zero but is 6050.25 in the original data (M3[[2851]]$xx[8] <- 6050.25)
  * Added the forecasts of the original M3 competition

## Version 2.04 (19 March 2012)
  * Changes to satisfy new requirements for R-2.15.0.

## Version 2.03 (11 August 2011)
	* Fixed error in plot.Mdata()

## Version 2.02 (23 February 2011)
  * Fixed errors in package help file

## Version 2.01 (19 October 2009)
  * Corrected URL errors in help files
	* Added option to subset.Mdata() to obtain the 111 series
	  used in the M-competition

## Version 2.00 (7 September 2009)
  * Package removed from forecasting bundle

For details of earlier versions, please see the Changelog of
the forecast package.



tests/testthat.R
library(testthat)
library(Mcomp)

test_check("Mcomp")


tests/testthat/test_Mcomp.R
context("Mcomp")

test_that("plot and autoplot return the right classes", {
  expect_equal(plot(M1[[1]]), NULL)
  expect_is(autoplot(M3[[1]]), "gg")
  expect_is(autoplot(M3[[1]]), "ggplot")
  expect_equal(length(autoplot(M3[[1]])), 9)
})

test_that("subset returns the right number of series using examples in help file", {
  expect_equal(length(subset(M3,4)), 756)
  expect_equal(length(subset(M1,1,"industry")), 35)
})



R/M1.R
#' M-Competition data
#'
#' The time series from the M1 forecasting competition.
#'
#' @format M1 is a list of 1001 series of class \code{Mcomp}.
#' Each series within \code{M1} is of class \code{Mdata} with the following structure:
#' \describe{
#'   \item{sn}{Name of the series}
#'   \item{st}{Series number and period. For example "Y1" denotes first yearly series,
#'   "Q20" denotes 20th quarterly series and so on.}
#'   \item{n}{The number of observations in the time series}
#'   \item{h}{The number of required forecasts}
#'   \item{period}{Interval of the time series.
#'             Possible values are "YEARLY", "QUARTERLY", "MONTHLY" & "OTHER".}
#'   \item{type}{The type of series.
#'             Possible values are "DEMOGR", "INDUST", "MACRO1", "MACRO2",
#'             "MICRO1", "MICRO2" & "MICRO3".}
#'   \item{description}{A short description of the time series}
#'   \item{x}{A time series of length \code{n} (the historical data)}
#'   \item{xx}{A time series of length \code{h} (the future data)}
#' }
#' @author Muhammad Akram and Rob Hyndman
#' @seealso \code{\link{M3}}, \code{\link{subset.Mcomp}}, \code{\link{plot.Mdata}}
#' @references Makridakis, S., A. Andersen, R. Carbone, R. Fildes, M. Hibon, R.
#' Lewandowski, J. Newton, E. Parzen, and R. Winkler (1982) The accuracy of
#' extrapolation (time series) methods: results of a forecasting competition.
#' \emph{Journal of Forecasting}, \bold{1}, 111--153.
#'
#' @source
#' \url{http://forecasters.org/resources/time-series-data/m-competition/}.
#'
#' @keywords datasets
#' @examples
#'
#' M1
#' plot(M1$YAF2)
#' subset(M1,"monthly")
"M1"


R/M3.R
#' M3-Competition data
#'
#' The time series from the M3 forecasting competition.
#'
#' @format M3 is a list of 3003 series of class \code{Mcomp}. Each series within \code{M3} is
#' of class \code{Mdata} with the following structure:
#' \describe{
#'   \item{sn}{Name of the series}
#'   \item{st}{Series number and period. For example "Y1" denotes first yearly series,
#'   "Q20" denotes 20th quarterly series and so on.}
#'   \item{n}{The number of observations in the time series}
#'   \item{h}{The number of required forecasts}
#'   \item{period}{Interval of the time series.
#'             Possible values are "YEARLY", "QUARTERLY", "MONTHLY" & "OTHER".}
#'   \item{type}{The type of series.
#'             Possible values for M3 are "DEMOGRAPHIC", "FINANCE", "INDUSTRY",
#'             "MACRO", "MICRO", "OTHER".}
#'   \item{description}{A short description of the time series}
#'   \item{x}{A time series of length \code{n} (the historical data)}
#'   \item{xx}{A time series of length \code{h} (the future data)}
#' }
#' @author Muhammad Akram and Rob Hyndman
#' @seealso \code{\link{subset.Mcomp}}, \code{\link{plot.Mdata}}
#' @references Makridakis and Hibon (2000) The M3-competition: results, conclusions and
#' implications. \emph{International Journal of Forecasting}, \bold{16},
#' 451-476.
#' @source
#' \url{http://forecasters.org/resources/time-series-data/m3-competition/}.
#'
#' @keywords datasets
#' @examples
#'
#' M3
#' plot(M3[[32]])
#' subset(M3,"monthly")
"M3"


R/plots.R
#' Plotting M Competition data
#'
#' Functions to plot a time series from the M competition data sets, showing
#' both the training and test sections of the series.
#'
#'
#' @aliases plot.Mdata autoplot.Mdata
#' @param x,object A series of M-competition data
#' @param xlim Limits on x-axis
#' @param ylim Limits on y-axis
#' @param main Main title
#' @param xlab Label on x-axis
#' @param ylab Label on y-axis
#' @param ... Other plotting arguments passed to \code{plot}. Ignored for
#' \code{autoplot}.
#' @return \code{autoplot.Mdata} returns a ggplot2 object, while
#' \code{plot.Mdata} returns nothing. Both functions produce a time series plot
#' of the selected series.
#' @author Rob Hyndman
#' @seealso \code{\link{M1}}, \code{\link{M3}}
#' @keywords hplot
#' @examples
#'
#' library(ggplot2)
#' plot(M1[[1]])
#' autoplot(M1$YAF3)
#' autoplot(M3[["N0647"]])
#' @export
#' @export plot.Mdata
plot.Mdata <- function(x, xlim=c(tsp(x$x)[1],tsp(x$xx)[2]), ylim=range(x$x,x$xx),
                       main=x$sn, xlab, ylab="", ...)
{
  freq <- frequency(x$x)
  if(missing(xlab))
  {
    if(start(x$x)[1] > 1900)
      xlab <- "Year"
    else
      xlab <- "Time"
  }
  plot(ts(c(x$x, rep(NA, x$h)), end = tsp(x$x)[2] + x$h/freq, frequency = freq),
    ylim=ylim, xlim=xlim, ylab=ylab, xlab=xlab, main=main, ...)
  lines(ts(x$xx, start = tsp(x$x)[2] + 1/freq, frequency = freq),lt=1,col=2)
}

#' @rdname plot.Mdata
#' @export
autoplot.Mdata <- function(object, ...)
{
  freq <- frequency(object$x)
  df <- cbind(
    Training=ts(c(object$x, rep(NA, object$h)), end = tsp(object$x)[2] + object$h/freq, frequency = freq),
    Test=ts(object$xx, start = tsp(object$x)[2] + 1/freq, frequency = freq))
  p <- autoplot(df) + ggtitle(object$sn) + ylab("")
  if(start(object$x)[1] > 1900)
    p <- p + xlab("Year")
  # Set up colours: black = training, red = test
  p <- p + scale_color_manual(values=c("black","red"), name="")
  return(p)
}



R/M3Forecast.R
#' M3-Competition forecasts of original competition participants
#'
#' The forecasts from all the original participating methods in the M3
#' forecasting competition.
#'
#' @format M3Forecast is a list of data.frames. Each list element is the result
#' of one forecasting method. The data.frame then has the following structure:
#' Each row is the forecast of one series. Rows are named accordingly. In total
#' there are 18 columns, i.e., 18 forecasts. If fewer forecasts than 18 exist,
#' the row is filled up with NA values.
#' @author Christoph Bergmeir and Rob Hyndman
#' @references Makridakis and Hibon (2000) The M3-competition: results,
#' conclusions and implications. \emph{International Journal of Forecasting},
#' \bold{16}, 451-476.
#' @source
#' \url{http://forecasters.org/resources/time-series-data/m3-competition/}.
#'
#' @keywords datasets
#' @examples
#'
#' M3Forecast[["NAIVE2"]][1,]
#'
#' \dontrun{
#' # calculate errors using the accuracy function
#' # from the forecast package
#'
#' errors <- lapply(M3Forecast, function(f) {
#'       res <- NULL
#'       for(x in 1:length(M3)) {
#'         curr_f <- unlist(f[x,])
#'         if(any(!is.na(curr_f))) {
#'           curr_res <- accuracy(curr_f, M3[[x]]$xx)
#'         } else {
#'           # if no results are available create NA results
#'           curr_res <- accuracy(M3[[x]]$xx, M3[[x]]$xx)
#'           curr_res <- rep(NA, length(curr_res))
#'         }
#'         res <- rbind(res, curr_res)
#'       }
#'       rownames(res) <- NULL
#'       res
#'     })
#'
#' ind_yearly <- which(unlist(lapply(M3, function(x) {x$period == "YEARLY"})))
#' ind_quarterly <- which(unlist(lapply(M3, function(x) {x$period == "QUARTERLY"})))
#' ind_monthly <- which(unlist(lapply(M3, function(x) {x$period == "MONTHLY"})))
#' ind_other <- which(unlist(lapply(M3, function(x) {x$period == "OTHER"})))
#'
#' yearly_errors <- t(as.data.frame(lapply(errors, function(x) {colMeans(x[ind_yearly,])})))
#' quarterly_errors <- t(as.data.frame(lapply(errors, function(x) {colMeans(x[ind_quarterly,])})))
#' monthly_errors <- t(as.data.frame(lapply(errors, function(x) {colMeans(x[ind_monthly,])})))
#' other_errors <- t(as.data.frame(lapply(errors, function(x) {colMeans(x[ind_other,])})))
#'
#' yearly_errors
#' quarterly_errors
#' monthly_errors
#' other_errors
#' }
"M3Forecast"


R/Mcomp-package.R
#' @importFrom graphics lines plot title
#' @importFrom stats frequency ts tsp start
#' @importFrom ggplot2 autoplot ggtitle xlab ylab scale_color_manual
#' @import forecast
NULL

#' Data from the M-competitions
#'
#' The 1001 time series from the M-competition (Makridakis et al. 1982), and
#' the 3003 time series and forecasts from the IJF-M3 competition (Makridakis
#' and Hibon, 2000).
#'
#'
#' @name Mcomp-package
#' @aliases Mcomp-package Mcomp
#' @docType package
#' @author Rob J Hyndman. \email{Rob.Hyndman@@monash.edu}, with assistance from
#' Muhammad Akram and Christoph Bergmeir.
#' @references Makridakis, S., A. Andersen, R. Carbone, R. Fildes, M. Hibon, R.
#' Lewandowski, J. Newton, E. Parzen, and R. Winkler (1982) The accuracy of
#' extrapolation (time series) methods: results of a forecasting competition.
#' \emph{Journal of Forecasting}, \bold{1}, 111--153.
#'
#' Makridakis and Hibon (2000) The M3-competition: results, conclusions and
#' implications. \emph{International Journal of Forecasting}, \bold{16},
#' 451-476.
#' @source
#' \url{http://forecasters.org/resources/time-series-data/m3-competition/}.
#'
#' @keywords package
NULL





R/Mdata.R
#' @export 
print.Mcomp <- function(x,...)
{
    n <- length(x)
    Type <- Period <- character(n)
    for(i in 1:n)
    {
        Type[i] <- x[[i]]$type
        if(Type[i]=="INDUSTRIAL")
            Type[i] <- "INDUST"
        if(strsplit(Type[i],"-")=="DEMOGRAPHI")
            Type[i] <- "DEMOGRAPHIC"
        Period[i] <- x[[i]]$period
    }
    if(length(unique(Type))==1)
        ty <- Type[1]
    if(length(unique(Period))==1)
        ty <- Period[1]

    if(length(unique(Type))>1 && length(unique(Period))>1)
    {
        cat(paste("M-Competition data:",n,"time series"),"\n\n")
        tab1 <- table(Period,Type)
        csum <- margin.table(tab1,2)
        tab2 <- as.table(rbind(tab1,Total=csum))
        rsum <- margin.table(tab2,1)
        tab3 <- as.table(cbind(tab2,Total=rsum),dnn=c("Period","Type of data"))
        names(dimnames(tab3))  <-  c("Period","Type of data")
        print(tab3)
    }
    else if(length(unique(Type))==1 && length(unique(Period))==1)
        print(table(Period,Type,dnn=c("Period","Type of data")))
    else
    {
        cat(paste("M-Competition data:",n,ty,"time series"),sep="","\n\n")
        print(table(Period,Type,dnn=c("Period","Type of data")))
    }
}

#' @export 
print.Mdata <- function(x,...)
{
    cat(paste(x$st),fill=TRUE,labels=paste("Series:"),sep="\n")
    cat(paste(x$type),fill=TRUE,labels=paste("Type of series:"),sep="\n")
    cat(paste(x$period),fill=TRUE,labels=paste("Period of series:"),sep="\n")
    cat(paste(x$description),fill=TRUE,labels=paste("Series description:"),sep="\n")
    cat("\nHISTORICAL data\n")
    print(x$x)
    cat("\nFUTURE data\n")
    print(x$xx)
}

Mcomp.sub <- function(x,getdata)
{
    n <- length(x)
    Type <- Period <- character(n)
    for(i in 1:n)
    {
        Type[i] <- x[[i]]$type
        if(Type[i]=="INDUSTRIAL")
            Type[i] <- "INDUST"
        if(strsplit(Type[i],"-")=="DEMOGRAPHI")
            Type[i] <- "DEMOGRAPHIC"
        Period[i] <- x[[i]]$period
    }
    getdatatable <- c("yearly","quarterly","monthly","other","macro","micro","industry",
                "finance","demographic","allother","macro1","macro2","micro1","micro2","micro3")
    if(is.character(getdata))
    {
        getdata <- getdatatable[charmatch(getdata,getdatatable)]
        if(length(getdata) != 1)
            stop("Ambiguous data type")
        else if(is.na(getdata))
            stop("Unknown data type")
    }

    if(getdata==1 | getdata=="yearly")
        choose <- (Period=="YEARLY")
    else if(getdata==4 | getdata=="quarterly")
        choose <- (Period=="QUARTERLY")
    else if(getdata==12 | getdata=="monthly")
        choose <- (Period=="MONTHLY")
    else if(getdata==111)
    {
        j <- match(x111,names(x))
        choose <- rep(FALSE,length(x))
        choose[j] <- TRUE
    }
    else if(getdata=="other")
        choose <-
        (Period=="OTHER")
    else if(getdata=="macro")
        choose <- (Type=="MACRO")
    else if(getdata=="micro")
        choose <- (Type=="MICRO")
    else if(getdata=="industry")
        choose <- (Type=="INDUSTRY" | Type=="INDUST")
    else if(getdata=="finance")
        choose <- (Type=="FINANCE")
    else if(getdata=="demographic")
        choose <- (Type=="DEMOGRAPHIC" | Type=="DEMOGR")
    else if(getdata=="allother")
        choose <- (Type=="OTHER")
    else if(getdata=="macro1")
        choose <- (Type=="MACRO1")
    else if(getdata=="macro2")
        choose <- (Type=="MACRO2")
    else if(getdata=="micro1")
        choose <- (Type=="MICRO1")
    else if(getdata=="micro2")
        choose <- (Type=="MICRO2")
    else if(getdata=="micro3")
        choose <- (Type=="MICRO3")
    else
        stop("Unknown type or period")
    if(sum(choose) == 0)
        stop("No data")

    return(x[choose])
}




#' Subset of time series from the M Competitions
#'
#' \code{subset.Mcomp} returns a subset of the time series data from the M
#' Competitions. Subsets can be for specific periods, or specific types of data
#' or both.
#'
#' Possible values for \code{cond1} and \code{cond2} denoting period are 1, 4,
#' 12, "yearly", "quarterly", "monthly" and "other".
#'
#' If \code{cond1} or \code{cond2} equals 111, then the 111 series used in the
#' extended comparisons in the 1982 M-competition are selected.
#'
#' Possible values for \code{cond1} and \code{cond2} denoting type are "macro",
#' "micro", "industry", "finance", "demographic", "allother", "macro1",
#' "macro2", "micro1", "micro2", "micro3". These correspond to the descriptions
#' used in the competitions. See the references for details.
#'
#' Partial matching used for both conditions.
#'
#' @param x M-competition data or a subset of M-competition data
#' @param cond1 Type or period of the data. Type is a character variable and
#' period could be character or numeric.
#' @param cond2 Optional second condition specifying type or period of the
#' data, depending on \code{cond1}.  If \code{cond1} denotes type then
#' \code{cond2} would denote period, but if \code{cond1} denotes period then
#' \code{cond2} would denote type.
#' @param ... Other arguments.
#' @return An object of class \code{Mcomp} consisting of the selected series.
#' @author Muhammad Akram and Rob Hyndman
#' @seealso \code{\link{M1}}
#' @references Makridakis, S., A. Andersen, R. Carbone, R. Fildes, M. Hibon, R.
#' Lewandowski, J. Newton, E. Parzen, and R. Winkler (1982) The accuracy of
#' extrapolation (time series) methods: results of a forecasting competition.
#' \emph{Journal of Forecasting}, \bold{1}, 111--153.
#'
#' Makridakis and Hibon (2000) The M3-competition: results, conclusions and
#' implications. \emph{International Journal of Forecasting}, \bold{16},
#' 451-476.
#' @keywords data
#' @examples
#'
#' M3.quarterly <- subset(M3,4)
#' M1.yearly.industry <- subset(M1,1,"industry")
#' @export
#' @export subset.Mcomp
#'
subset.Mcomp <- function(x,cond1,cond2,...)
{
    if(is.character(cond1))
        cond1 <- tolower(cond1)
    M11=structure(Mcomp.sub(x,cond1),class="Mcomp")
    if(!missing(cond2))
    {
        if(is.character(cond2))
            cond2 <- tolower(cond2)
        M12=structure(Mcomp.sub(M11,cond2),class="Mcomp")
        return(M12)
    }
    else
        return(M11)
}


#' @export
`[.Mcomp` <- function(x, i)
{
  y <- NextMethod("[")
  class(y) <- c("Mcomp", class(y))
  y
}
